src/tools/chip-cert/Cmd_PrintCD.cpp

```diff
@@ -0,0 +1,277 @@
+
+// clang-format off
+OptionDef gCmdOptionDefs[] =
+{
+    { "out",           kArgumentRequired, 'o' },
+    { }
+};
+
+const char * const gCmdOptionHelp =
+    "   -o, --out <file/stdout>\n"
+    "\n"
+    "       The output printed CD content file name. If not specified\n"
+    "       or if specified '-' then output is written to stdout.\n"
+    "\n"
+    ;
+
+OptionSet gCmdOptions =
+{
+    HandleOption,
+    gCmdOptionDefs,
+    "COMMAND OPTIONS",
+    gCmdOptionHelp
+};
+
+HelpOptions gHelpOptions(
+    CMD_NAME,
+    "Usage: " CMD_NAME " [<options...>] <file/str>\n",
+    CHIP_VERSION_STRING "\n" COPYRIGHT_STRING,
+    "Print a CHIP certification declaration (CD) content.\n"
+    "\n"
+    "ARGUMENTS\n"
+    "\n"
+    "  <file/str>\n"
+    "\n"
+    "       File or string containing a CHIP CMS Signed CD message.\n"
+    "\n"
+);
+
+OptionSet *gCmdOptionSets[] =
+{
+    &gCmdOptions,
+    &gHelpOptions,
+    nullptr
+};
+// clang-format on
+
+const char * gInFileNameOrStr = nullptr;
+const char * gOutFileName     = "-";
+
+bool HandleOption(const char * progName, OptionSet * optSet, int id, const char * name, const char * arg)
+{
+    switch (id)
+    {
+    case 'o':
+        gOutFileName = arg;
+        break;
+    default:
+        PrintArgError("%s: Unhandled option: %s\n", progName, name);
+        return false;
+    }
+
+    return true;
+}
+
+bool HandleNonOptionArgs(const char * progName, int argc, char * const argv[])
+{
+    if (argc == 0)
+    {
+        PrintArgError("%s: Please specify the CD to be printed.\n", progName);
+        return false;
+    }
+
+    if (argc > 1)
+    {
+        PrintArgError("%s: Unexpected argument: %s\n", progName, argv[1]);
+        return false;
+    }
+
+    gInFileNameOrStr = argv[0];
+
+    return true;
+}
+
+enum CDFormat
+{
+    kCDFormat_Unknown = 0,
+    kCDFormat_Raw,
+    kCDFormat_Hex,
+    kCDFormat_Base64,
+};
+
+CDFormat DetectCDFormat(const uint8_t * cd, uint32_t cdLen)
+{
+    static const uint8_t cdRawPrefix1[] = { 0x30, 0x81 };
+    static const uint8_t cdRawPrefix2[] = { 0x30, 0x82 };
+    static const char * cdHexPrefix     = "308";
+    static const char * cdB64Prefix     = "MI";
+
+    VerifyOrReturnError(cd != nullptr, kCDFormat_Unknown);
+
+    if ((cdLen > sizeof(cdRawPrefix1)) &&
+        (memcmp(cd, cdRawPrefix1, sizeof(cdRawPrefix1)) == 0 || memcmp(cd, cdRawPrefix2, sizeof(cdRawPrefix2)) == 0))
+    {
+        return kCDFormat_Raw;
+    }
+
+    if ((cdLen > strlen(cdHexPrefix)) && (memcmp(cd, cdHexPrefix, strlen(cdHexPrefix)) == 0))
+    {
+        return kCDFormat_Hex;
+    }
+
+    if ((cdLen > strlen(cdB64Prefix)) && (memcmp(cd, cdB64Prefix, strlen(cdB64Prefix)) == 0))
+    {
+        return kCDFormat_Base64;
+    }
+
+    return kCDFormat_Unknown;
+}
+
+bool ReadCD(const char * fileNameOrStr, MutableByteSpan cd)
+{
+    CDFormat cdFmt = kCDFormat_Unknown;
+    uint32_t cdLen = 0;
+    std::unique_ptr<uint8_t[]> cdBuf;
```

Comment:
A better object here is `std::vector`

Reply:
Good Point! I'll do that change as a followup as it requires changes in other places in this tool.